backend/models/upload.js

```diff
@@ -0,0 +1,240 @@
+	);
+
+	return { corID };
+};
+
+Upload.uploadFile = async (req) => {
+	if (!config.cn_queue) {
+		return Promise.reject(responseCodes.QUEUE_NO_CONFIG);
+	}
+
+	const { account, model } = req.params;
+	const user = req.session.user.username;
+
+	modelStatusChanged(null, account, model, { status: "uploading", user });
+	// upload model with tag
+
+	const uploadedFile = await new Promise((resolve, reject) => {
+		const upload = multer({
+			dest: config.cn_queue.upload_dir,
+			fileFilter: async function(fileReq, file, cb) {
+				const size = parseInt(fileReq.headers["content-length"]);
+
+				try {
+					await checkFileFormat(file.originalname);
+					await middlewares.checkSufficientSpace(account, size);
+					cb(null, true);
+				} catch (err) {
+					cb(err);
+				}
+			}
+		});
+
+		upload.single("file")(req, null, function (err) {
+			if (err) {
+				return reject(err);
+
+			} else if(!req.file.size) {
+				return reject(responseCodes.FILE_FORMAT_NOT_SUPPORTED);
+
+			} else {
+				modelStatusChanged(null, account, model, { status: "uploaded" });
+				return resolve(req.file);
+			}
+		});
+	});
+
+	// req.body.tag wont be defined after the file has been uploaded
+	await isValidTag(account, model, req.body.tag);
+
+	return uploadedFile;
+};
+
+Upload.uploadChunksStart = async (teamspace, model, corID, username, headers) => {
+	if (!headers["x-ms-transfer-mode"] ||
+		headers["x-ms-transfer-mode"] !== "chunked" ||
+		!headers["x-ms-content-length"] ||
+		isNaN(headers["x-ms-content-length"])) {
+		// for debugging
+		systemLogger.logInfo(`transfer mode=${headers["x-ms-transfer-mode"]}`);
+		systemLogger.logInfo(`content length=${headers["x-ms-content-length"]}`);
+		// end debugging
+		throw responseCodes.INVALID_ARGUMENTS;
+	}
+
+	await middlewares.checkSufficientSpace(teamspace, parseInt(headers["x-ms-content-length"]));
+	const chunkSize = Math.min(C.MS_CHUNK_BYTES_LIMIT, parseInt(headers["x-ms-content-length"]));
+
+	if (!fs.existsSync(`${importQueue.getTaskPath(corID)}.json`)) {
+		throw responseCodes.CORRELATION_ID_NOT_FOUND;
+	}
+
+	// upload model with tag
+	modelStatusChanged(null, teamspace, model, { status: "uploading", username });
+
+	await utils.mkdir(`${importQueue.getTaskPath(corID)}/chunks/`);
+
+	return { "x-ms-chunk-size": chunkSize };
+};
+
+Upload.uploadChunk = async (teamspace, model, corID, req) => {
+	if (!fs.existsSync(`${importQueue.getTaskPath(corID)}.json`)) {
+		throw responseCodes.CORRELATION_ID_NOT_FOUND;
+	}
+
+	if (!config.cn_queue) {
+		return Promise.reject(responseCodes.QUEUE_NO_CONFIG);
+	}
+
+	if (!req.headers["content-range"]) {
+		throw responseCodes.INVALID_ARGUMENTS;
+	}
+
+	const [contentRange, totalContentSize] = req.headers["content-range"].split("/");
+	const [sizeUnit, contentRangeValue] = contentRange.split(" ");
+
+	if (sizeUnit !== "bytes") {
+		throw responseCodes.INVALID_ARGUMENTS;
+	}
+
+	await middlewares.checkSufficientSpace(teamspace, parseInt(totalContentSize));
```

Comment:
does this mean you're checking it everytime there's a chunk coming in?

I think we only need to check it on the last chunk. it's a bit excessive to do it at every chunk.